test/runtests.jl

```diff
@@ -1,95 +1,193 @@
+    ("BGRN0f8", identity, rand(BGR{N0f8}, 127, 257)),
+    ("BGRAN0f8", identity, rand(BGRA{N0f8}, 127, 257)),
+    ("BGRN0f16", identity, rand(BGR{N0f16}, 127, 257)),
+    ("BGRAN0f16", identity, rand(BGRA{N0f16}, 127, 257)),
+    ("ABGRN0f8", identity, rand(ABGR{N0f8}, 127, 257)),
+    ("ABGRN0f16", identity, rand(ABGR{N0f16}, 127, 257)),
+    ("ARGBN0f8", identity, rand(ARGB{N0f8}, 127, 257)),
+    ("ARGBN0f16", identity, rand(ARGB{N0f16}, 127, 257)),
+]
+
+@testset "libpng" begin
+    for (case, image) in vcat(synth_imgs, real_imgs)
+        @debug case
+        @testset "$(case)" begin
+            expected = collect(_prepare_buffer(image))
+            filename = File{DataFormat{:PNG}}(joinpath(PNG_TEST_PATH, "test_img_$(case).png"))
+            @testset "write" begin
+                @test save(filename, image) == 0
+            end
+            @testset "read" begin
+                global read_in = load(filename)
+                @test read_in isa Matrix
+            end
+            @testset "compare" begin
+                @test all(expected .≈ read_in)
+            end
+        end
     end
 
-    @testset "Gray image" begin
-        gray = vcat(fill(Gray(1.0), 10, 10), fill(Gray(0.5), 10, 10), fill(Gray(0.0), 10, 10))
-        filepath = joinpath(tmpdir, "gray1.png")
-        PNG.save(File{DataFormat{:PNG}}(filepath), gray)
-        g1 = PNG.load(File{DataFormat{:PNG}}(filepath))
-        @test g1 == convert(Array{Gray{N0f8}}, gray)
-
-        gray = rand(Gray{N0f8}, 5, 5)
-        filepath = joinpath(tmpdir, "gray2.png")
-        PNG.save(File{DataFormat{:PNG}}(filepath), gray)
-        g2 = PNG.load(File{DataFormat{:PNG}}(filepath))
-        @test g2 == gray
+    for (case, exception, image) in invalid_imgs
+        @debug case
+        @testset "$(case) throws" begin
+            @test_throws exception save(
+                File{DataFormat{:PNG}}(joinpath(PNG_TEST_PATH, "test_img_err_$(case).png")),
+                image
+            )
+        end
     end
 
-    @testset "Color - RGB" begin
-        #rgb8 = rand(RGB{N0f8}, 10, 5)
-        rgb8 = reshape(range(RGB{N0f8}(1,0,0),RGB{N0f8}(0,1,1), length=10*5), 10, 5)
-        filepath = joinpath(tmpdir, "rgb_n0f8.png")
-        PNG.save(File{DataFormat{:PNG}}(filepath), rgb8)
-        r1 = PNG.load(File{DataFormat{:PNG}}(filepath))
-        @test r1 == rgb8
-
-        #rgb16 = rand(RGB{N0f16}, 10, 5)
-        rgb16 = reshape(range(RGB{N0f16}(1,0,0),RGB{N0f16}(0,1,1), length=10*5), 10, 5)
-        filepath = joinpath(tmpdir, "rgb_n0f16.png")
-        PNG.save(File{DataFormat{:PNG}}(filepath), rgb16)
-        r2 = PNG.load(File{DataFormat{:PNG}}(filepath))
-        PNG.save(File{DataFormat{:PNG}}(joinpath(tmpdir, "rgb_n0f16_resave.png")), r2)
-        @test r2 == rgb16
+    for (case, func_in, image) in edge_case_imgs
+        @debug case
+        @testset "$(case)" begin
+            filename = File{DataFormat{:PNG}}(joinpath(PNG_TEST_PATH, "test_img_$(case).png"))
```

Comment:
I think it would be clearer to refer to this as `f` as it's more than a filename

Reply:
I'm happy to make this change post merge (it's very nitpicky)

Reply:
:+1: